include/gatekeeper_gk.h

```diff
@@ -19,17 +19,126 @@
 #ifndef _GATEKEEPER_GK_H_
 #define _GATEKEEPER_GK_H_
 
+#include <stdint.h>
+
 #include <rte_atomic.h>
 
+#define	START_PRIORITY		(38)
+/* Set @START_ALLOWANCE as the double size of a large DNS reply. */
+#define	START_ALLOWANCE		(8)
+
+/* Structures for each GK instance. */
+struct gk_instance {
+	struct rte_hash   *ip_flow_hash_table;
+	struct flow_entry *ip_flow_entry_table;
+};
+
 /* Configuration for the GK functional block. */
 struct gk_config {
-	unsigned int	lcore_start_id;
-	unsigned int	lcore_end_id;
+	unsigned int	   lcore_start_id;
+	unsigned int	   lcore_end_id;
+
+	/* Specify the size of the flow hash table. */
+	unsigned int	   flow_ht_size;
+
 	/*
 	 * The fields below are for internal use.
 	 * Configuration files should not refer to them.
 	 */
-	rte_atomic32_t	ref_cnt;
+	rte_atomic32_t	   ref_cnt;
+	unsigned int	   num_lcores;
+	struct gk_instance *instances;
+};
+
+struct ip_flow {
+	/* IPv4 or IPv6. */
+	uint16_t proto;
+
+	union {
+		struct {
+			uint32_t src;
+			uint32_t dst;
+		} v4;
+
+		struct ipv6_flow {
+			uint8_t src[16];
+			uint8_t dst[16];
+		} v6;
+	} f;
+};
+
+/*
+ * XXX Stores information about the packet, add new fields if necessary.
+ */
+struct ipacket {
+	struct ip_flow  flow;
+	struct rte_mbuf *pkt;
+};
+
+struct flow_entry {
+	/* IP flow information. */
+	struct ip_flow flow;
+
+	/* The state of the entry. */
+	unsigned state : 2;
+
+	union {
+		struct {
+			/* The time the last packet of the entry was seen. */
+			uint64_t last_packet_seen_at;
+			/* 
+			 * The priority associated to
+			 * the last packet of the entry.
+			 */
+			uint8_t last_priority;
+			/* 
+			 * The number of packets that the entry is allowed
+			 * to send with @last_priority without waiting
+			 * the amount of time necessary to be granted
+			 * @last_priority.
+			 */
+			uint8_t allowance;
+			/* 
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+		} request;
+
+		struct {
+			/* When the granted capability expires. */
+			uint64_t cap_expire_at;
+			/* When @budget_byte is reset. */
+			uint64_t budget_renew_at;
+			/* 
+			 * When @budget_byte is reset, reset it to
+			 * @tx_rate_kb_cycle * 1024 bytes.
+			 */
+			int tx_rate_kb_cycle;
+			/* How many bytes @src can still send in current cycle. */
+			int budget_byte;
+			/*
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+			/* When GK should send the next renewal to @grantor_id. */
+			uint64_t send_next_renewal_at;
+			/*
+			 * How many cycles (unit) GK must wait before
+			 * sending the next capability renewal request.
+			 */
+			uint64_t renewal_step_cycle;
+		} granted;
+
+		struct {
+			/*
+			 * When the punishment (i.e. the declined capability)
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
```

Comment:
I think these three struct definitions can go in `gk/main.c`, since they are only used there.